### ISSUE TITLE
output of -w and -o are not consistent when run with -d 2

### ISSUE THREAD

**Issue by nhoffman** [torognes/swarm#67]
Hi,

It appears that the output of -w/seeds and -o/--output-file are not consistent with one another when I run swarm with -d 2:

d0.fasta is a file containing dereplicated reads.

```
% ./swarm --version
Swarm 2.1.5 [Sep  8 2015 14:33:53]
% ./swarm -d 1 -w d1.fasta -o d1_clusters.txt d0.fasta
% grep '>' d1.fasta | head -n3
>H7CGFRN01AMXNS_695
>H7CGFRN01A9B06_911
>H7CGFRN01DNP2R_1126
% cut -f 1-5 -d ' ' d1_clusters.txt | head -n 3
H7CGFRN01AMXNS_474 H7CGFRN01D22V9_35 H7CGFRN01BJ73R_14 H7CGFRN01DQM49_8 H7CGFRN01D99ZU_7
H7CGFRN01A9B06_410 H7CGFRN01APEVJ_136 H7CGFRN01CMNQO_47 H7CGFRN01B4OPI_10 H7CGFRN01DERYD_8
H7CGFRN01DNP2R_386 H7CGFRN01DMO0J_272 H7CGFRN01CZRK7_22 H7CGFRN01BKKEQ_19 H7CGFRN01AYHSA_9
``` 

So far, so good - we see that the seed sequence with the highest abundance (H7CGFRN01AMXNS) appears on the first line of each file, and the total abundance indicated for the first read of d1.fasta looks like the sum of the abundances in the first line of d1_clusters.txt.

Now for -d 2:

```
% ./swarm -d 2 -w d2.fasta -o d2_clusters.txt d0.fasta
% grep '>' d2.fasta | head -n3
>H7CGFRN01A5Q5F_1374
>H7CGFRN01APDGM_3103
>H7CGFRN01B7XJP_2384
% cut -f 1-5 -d ' ' d2_clusters.txt | head -n 3
H7CGFRN01AMXNS_474 H7CGFRN01CU01U_224 H7CGFRN01D22V9_35 H7CGFRN01BJ73R_14 H7CGFRN01DQM49_8
H7CGFRN01A9B06_410 H7CGFRN01CZ6LA_281 H7CGFRN01EAGGX_141 H7CGFRN01APEVJ_136 H7CGFRN01CMNQO_47
H7CGFRN01DNP2R_386 H7CGFRN01DMO0J_272 H7CGFRN01DEKVC_170 H7CGFRN01CZRK7_22 H7CGFRN01BKKEQ_19
```

I expected the first line of d2_clusters.txt to start with H7CGFRN01A5Q5F... is this the expected result? 

Here's a link to the input file: https://dl.dropboxusercontent.com/u/2673381/d0.fasta.gz

Thanks a lot for having a look.
Noah

**Comment by frederic-mahe** [comment]
Hi @nhoffman,

the problem is that `-w` output is not sorted by decreasing abundance. What you observe with `-d 2` can also occur with `-d 1` if you use another dataset. In my own pipeline, I've added a sorting step on the `-w` output with `vsearch`. I did not imagine someone else would need that file to be sorted.

Is it a correct interpretation of your bug report?

**Comment by nhoffman** [comment]
Hi @frederic-mahe,

Yes, this is a correct interpretation, and I certainly should have noticed that this is just a matter of sort order! I have a script that steps through the pair of files and processes each seed sequence along with the corresponding list of read names comprising the cluster. For what it's worth, the outputs of -o and -w  are consistently in the same order with -d 1 (at least for all of the data sets that I have tried). I assumed (incorrectly, it turns out) that this was by design, and that the difference in sort order between the files was a bug. It would be convenient to be able to assume that the output of -w and -o are in the same order, but now that I realize what's going on, not essential.

Thanks a lot,
Noah

**Comment by frederic-mahe** [comment]
Maybe we can add that as an enhancement:

sort OTU representatives by decreasing abundance (and by other features to stabilize sorting), or put them in the same order than in the swarm output file and stats output file.

**Comment by nhoffman** [comment]
Yes - even if the sort order is arbitrary, it would be useful for it to be the same among the various outputs.  Thanks for considering this!

**Comment by torognes** [comment]
Thanks for the suggestion. I will add this in the next version.

**Comment by torognes** [comment]
I am looking into this now, and it seems to be a bug. When using d>1, the seed id shown is the id of the last/smallest amplicon in the cluster, not the first/largest. Will fix very soon.

**Comment by torognes** [comment]
Thanks again for reporting this problem that appeared to be a real bug. Problem fixed in Swarm version 2.1.6.

**Comment by frederic-mahe** [comment]
Hi everyone,

Sorry for not catching that bug earlier. Most of the time, I use `d = 1` and I did not spend enough time checking the results of `d > 1 runs`. I will try to write short scripts to test bugs like this, to prevent regressions in future versions of swarm. The script below checks if the list of fasta representatives and the list of swarm seeds contain exactly the same elements. The script uses the fasta dataset provided by @nhoffman. It fails on `d > 1` with swarm 2.1.5, and it passes for all `d` values with swarm 2.1.6:
```sh
#!/bin/bash
cd ~/swarm/tests/bug_issue_67/

ORIGINAL="d0.fasta"
SWARMS=$(mktemp)
REPRESENTATIVES=$(mktemp)
UNIQUE=$(mktemp)

# Inform user
echo "## Test of issue 67"
echo "# list of fasta representatives should be the same than the list of swarm seeds"

# Check if data is here
if [[ ! -r "${ORIGINAL}.bz2" ]] ; then
    echo "file not readable"
    exit -1
else
    bunzip2 -k "${ORIGINAL}.bz2"
fi

# Loop over interesting d values
for i in {1..3} ; do
    echo -en "run swarm with d = ${i}\t"
    swarm -d ${i} -w "${REPRESENTATIVES}" "${ORIGINAL}" > "${SWARMS}" 2> /dev/null
    # Search for differences between lists of swarms and fasta representatives
    (grep "^>" "${REPRESENTATIVES}" | tr -d ">" | cut -d "_" -f 1 ; \
     cut -d "_" -f 1 "${SWARMS}") | sort -d | uniq -u > "${UNIQUE}"
    if [[ -s "${UNIQUE}" ]] ; then
        printf '\e[1;31m%-6s\e[m\n' "failed"
    else
        printf '\e[1;32m%-6s\e[m\n' "passed"
    fi
done

rm "${ORIGINAL}" "${REPRESENTATIVES}" "${SWARMS}" "${UNIQUE}"

exit 0
```

**Comment by torognes** [comment]
@nhoffman I am sorry but this problem was not properly fixed in version 2.1.6 from December. Only the id of the seed was corrected then, the actual sequence was still wrong. This problem was fixed yesterday in version 2.1.7. This bug only appears when d>1. Thanks again for pointing out the problem. I hope it has not caused major problems.

**Comment by nhoffman** [comment]
@torognes Thanks for letting me know. No problems on my end - I always use `d = 1` myself.
